qmail-queue.8
.TH qmail-queue 8
.SH NAME
qmail-queue \- queue a mail message for delivery
.SH SYNOPSIS
.B qmail-queue
.SH DESCRIPTION
.B qmail-queue
reads a mail message from descriptor 0.
It then reads envelope information from descriptor 1.
It places the message into the outgoing queue
for future delivery by
.BR qmail-send .

The envelope information is
an envelope sender address
followed by a list of envelope recipient addresses.
The sender address is preceded by the letter F
and terminated by a 0 byte.
Each recipient address is preceded by the letter T
and terminated by a 0 byte.
The list of recipient addresses is terminated by an extra 0 byte.
If
.B qmail-queue
sees end-of-file before the extra 0 byte,
it aborts without placing the message into the queue.

Every envelope recipient address
should contain a username,
an @ sign,
and a fully qualified domain name.

.B qmail-queue
always adds a
.B Received
line to the top of the message.
Other than this,
.B qmail-queue
does not inspect the message
and does not enforce any restrictions on its contents.
However, the recipients probably expect to see a proper header,
as described in
.BR qmail-header(5) .

Programs included with qmail which invoke
.B qmail-queue
will invoke the contents of
.B $QMAILQUEUE
instead, if that environment variable is set.
.SH "CONTROL FILES"
.TP 5
.I taps
Should contain source address (T for To,F for From,
A for Any), regex syntax of email addresses to tap 
and the associated email address to send the copy to.
The fields should be separated by colon.
.SH "FILESYSTEM RESTRICTIONS"
.B qmail-queue
imposes two constraints on the queue structure:
each
.B mess
subdirectory must be in the same filesystem as the
.B pid
directory; and each
.B todo
subdirectory must be in the same filesystem as the
.B intd
directory.
.SH "EXIT CODES"
.B qmail-queue
does not print diagnostics.
It exits
0 if
it has successfully queued the message.
It exits between 1 and 99 if
it has failed to queue the message.

All
.B qmail-queue
error codes between 11 and 40
indicate permanent errors:
.TP 5
.B 11
Address too long.
.TP
.B 31
Mail server permanently refuses to send the message to any recipients.
(Not used by
.BR qmail-queue ,
but can be used by programs offering the same interface.)
.PP
All other
.B qmail-queue
error codes indicate temporary errors:
.TP 5
.B 51
Out of memory.
.TP
.B 52
Timeout.
.TP
.B 53
Write error; e.g., disk full.
.TP
.B 54
Unable to read the message or envelope.
.TP
.B 55
Unable to read a configuration file.
(Not used by
.BR qmail-queue .)
.TP
.B 56
Problem making a network connection from this host.
(Not used by
.BR qmail-queue .)
.TP
.B 61
Problem with the qmail home directory.
.TP
.B 62
Problem with the queue directory.
.TP
.B 63
Problem with queue/pid.
.TP
.B 64
Problem with queue/mess.
.TP
.B 65
Problem with queue/intd.
.TP
.B 66
Problem with queue/todo.
.TP
.B 71
Mail server temporarily refuses to send the message to any recipients.
(Not used by
.BR qmail-queue .)
.TP
.B 72
Connection to mail server timed out.
(Not used by
.BR qmail-queue .)
.TP
.B 73
Connection to mail server rejected.
(Not used by
.BR qmail-queue .)
.TP
.B 74
Connection to mail server succeeded,
but communication failed.
(Not used by
.BR qmail-queue .)
.TP
.B 81
Internal bug; e.g., segmentation fault.
.TP
.B 91
Envelope format error.
.SH "SEE ALSO"
addresses(5),
envelopes(5),
qmail-header(5),
qmail-inject(8),
qmail-qmqpc(8),
qmail-send(8),
qmail-smtpd(8)


qmail-queue.c
#include <sys/types.h>
#include <sys/stat.h>
#include "readwrite.h"
#include "sig.h"
#include "exit.h"
#include "open.h"
#include "seek.h"
#include "fmt.h"
#include "alloc.h"
#include "substdio.h"
#include "datetime.h"
#include "now.h"
#include "triggerpull.h"
#include "extra.h"
#include "auto_qmail.h"
#include "auto_uids.h"
#include "date822fmt.h"
#include "fmtqfn.h"
#include "stralloc.h"
#include "constmap.h"

#define DEATH 86400 /* 24 hours; _must_ be below q-s's OSSIFIED (36 hours) */
#define ADDR 1003

char inbuf[2048];
struct substdio ssin;
char outbuf[256];
struct substdio ssout;

int tapok = 0;
stralloc tap = {0};
struct constmap maptap;
stralloc chkaddr = {0};
int tapped;
stralloc tapaddr = {0};
stralloc controlfile = {0};

datetime_sec starttime;
struct datetime dt;
unsigned long mypid;
unsigned long uid;
char *pidfn;
struct stat pidst;
unsigned long messnum;
char *messfn;
char *todofn;
char *intdfn;
int messfd;
int intdfd;
int flagmademess = 0;
int flagmadeintd = 0;

void cleanup()
{
 if (flagmadeintd)
  {
   seek_trunc(intdfd,0);
   if (unlink(intdfn) == -1) return;
  }
 if (flagmademess)
  {
   seek_trunc(messfd,0);
   if (unlink(messfn) == -1) return;
  }
}

void die(e) int e; { _exit(e); }
void die_write() { cleanup(); die(53); }
void die_read() { cleanup(); die(54); }
void sigalrm() { /* thou shalt not clean up here */ die(52); }
void sigbug() { die(81); }

unsigned int receivedlen;
char *received;
/* "Received: (qmail-queue invoked by alias); 26 Sep 1995 04:46:54 -0000\n" */

static unsigned int receivedfmt(s)
char *s;
{
 unsigned int i;
 unsigned int len;
 len = 0;
 i = fmt_str(s,"Received: (qmail "); len += i; if (s) s += i;
 i = fmt_ulong(s,mypid); len += i; if (s) s += i;
 i = fmt_str(s," invoked "); len += i; if (s) s += i;
 if (uid == auto_uida)
  { i = fmt_str(s,"by alias"); len += i; if (s) s += i; }
 else if (uid == auto_uidd)
  { i = fmt_str(s,"from network"); len += i; if (s) s += i; }
 else if (uid == auto_uids)
  { i = fmt_str(s,"for bounce"); len += i; if (s) s += i; }
 else
  {
   i = fmt_str(s,"by uid "); len += i; if (s) s += i;
   i = fmt_ulong(s,uid); len += i; if (s) s += i;
  }
 i = fmt_str(s,"); "); len += i; if (s) s += i;
 i = date822fmt(s,&dt); len += i; if (s) s += i;
 return len;
}

void received_setup()
{
 receivedlen = receivedfmt((char *) 0);
 received = alloc(receivedlen + 1);
 if (!received) die(51);
 receivedfmt(received);
}

unsigned int pidfmt(s,seq)
char *s;
unsigned long seq;
{
 unsigned int i;
 unsigned int len;

 len = 0;
 i = fmt_str(s,"pid/"); len += i; if (s) s += i;
 i = fmt_ulong(s,mypid); len += i; if (s) s += i;
 i = fmt_str(s,"."); len += i; if (s) s += i;
 i = fmt_ulong(s,starttime); len += i; if (s) s += i;
 i = fmt_str(s,"."); len += i; if (s) s += i;
 i = fmt_ulong(s,seq); len += i; if (s) s += i;
 ++len; if (s) *s++ = 0;

 return len;
}

char *fnnum(dirslash,flagsplit)
char *dirslash;
int flagsplit;
{
 char *s;

 s = alloc(fmtqfn((char *) 0,dirslash,messnum,flagsplit));
 if (!s) die(51);
 fmtqfn(s,dirslash,messnum,flagsplit);
 return s;
}

void pidopen()
{
 unsigned int len;
 unsigned long seq;

 seq = 1;
 len = pidfmt((char *) 0,seq);
 pidfn = alloc(len);
 if (!pidfn) die(51);

 for (seq = 1;seq < 10;++seq)
  {
   if (pidfmt((char *) 0,seq) > len) die(81); /* paranoia */
   pidfmt(pidfn,seq);
   messfd = open_excl(pidfn);
   if (messfd != -1) return;
  }

 die(63);
}

char tmp[FMT_ULONG];

void main()
{
 unsigned int len;
 char ch;

 sig_blocknone();
 umask(033);
 if (chdir(auto_qmail) == -1) die(61);
 if (chdir("queue") == -1) die(62);

 mypid = getpid();
 uid = getuid();
 starttime = now();
 datetime_tai(&dt,starttime);

 received_setup();

 sig_pipeignore();
 sig_miscignore();
 sig_alarmcatch(sigalrm);
 sig_bugcatch(sigbug);

 alarm(DEATH);

 stralloc_copys( &controlfile, auto_qmail);
 stralloc_cats( &controlfile, "/control/taps");
 stralloc_0( &controlfile);
 tapok = control_readfile(&tap,controlfile.s,0);
 if (tapok == -1) die(65);
 if (!constmap_init(&maptap,tap.s,tap.len,0)) die(65);

 pidopen();
 if (fstat(messfd,&pidst) == -1) die(63);

 messnum = pidst.st_ino;
 messfn = fnnum("mess/",1);
 todofn = fnnum("todo/",1);
 intdfn = fnnum("intd/",1);

 if (link(pidfn,messfn) == -1) die(64);
 if (unlink(pidfn) == -1) die(63);
 flagmademess = 1;

 substdio_fdbuf(&ssout,write,messfd,outbuf,sizeof(outbuf));
 substdio_fdbuf(&ssin,read,0,inbuf,sizeof(inbuf));

 if (substdio_bput(&ssout,received,receivedlen) == -1) die_write();

 switch(substdio_copy(&ssout,&ssin))
  {
   case -2: die_read();
   case -3: die_write();
  }

 if (substdio_flush(&ssout) == -1) die_write();
 if (fsync(messfd) == -1) die_write();

 intdfd = open_excl(intdfn);
 if (intdfd == -1) die(65);
 flagmadeintd = 1;

 substdio_fdbuf(&ssout,write,intdfd,outbuf,sizeof(outbuf));
 substdio_fdbuf(&ssin,read,1,inbuf,sizeof(inbuf));

 if (substdio_bput(&ssout,"u",1) == -1) die_write();
 if (substdio_bput(&ssout,tmp,fmt_ulong(tmp,uid)) == -1) die_write();
 if (substdio_bput(&ssout,"",1) == -1) die_write();

 if (substdio_bput(&ssout,"p",1) == -1) die_write();
 if (substdio_bput(&ssout,tmp,fmt_ulong(tmp,mypid)) == -1) die_write();
 if (substdio_bput(&ssout,"",1) == -1) die_write();

 if (substdio_get(&ssin,&ch,1) < 1) die_read();
 if (ch != 'F') die(91);
 if (substdio_bput(&ssout,&ch,1) == -1) die_write();
 stralloc_0(&chkaddr);
 for (len = 0;len < ADDR;++len)
  {
   if ( len == 1 ) stralloc_copyb(&chkaddr, &ch,1);
   else if ( len > 1 ) stralloc_catb(&chkaddr, &ch,1);
   if (substdio_get(&ssin,&ch,1) < 1) die_read();
   if (substdio_put(&ssout,&ch,1) == -1) die_write();
   if (!ch) break;
  }
 if (len >= ADDR) die(11);

 /* check the from address */
 stralloc_0(&chkaddr);
 if (tapped == 0 && tapcheck('F')==1 ) {
   tapped = 1;
   if ( tapaddr.len > 0 ) {
     if (substdio_bput(&ssout,"T",1) == -1) die_write();
     if (substdio_bput(&ssout,tapaddr.s,tapaddr.len) == -1) die_write();
     if (substdio_bput(&ssout,"",1) == -1) die_write();
   }
 }

 if (substdio_bput(&ssout,QUEUE_EXTRA,QUEUE_EXTRALEN) == -1) die_write();

 for (;;)
  {
   if (substdio_get(&ssin,&ch,1) < 1) die_read();
   if (!ch) break;
   if (ch != 'T') die(91);
   if (substdio_bput(&ssout,&ch,1) == -1) die_write();
   for (len = 0;len < ADDR;++len)
    {
     if ( len == 1 ) stralloc_copyb(&chkaddr, &ch,1);
     else if ( len > 1 ) stralloc_catb(&chkaddr, &ch,1);
     if (substdio_get(&ssin,&ch,1) < 1) die_read();
     if (substdio_bput(&ssout,&ch,1) == -1) die_write();
     if (!ch) break;
    }

    /* check the to address */
    stralloc_0(&chkaddr);
    if (tapped == 0 && tapcheck('T')==1 ) {
      tapped = 1;
      if ( tapaddr.len > 0 ) {
        if (substdio_bput(&ssout,"T",1) == -1) die_write();
        if (substdio_bput(&ssout,tapaddr.s,tapaddr.len) == -1) die_write();
        if (substdio_bput(&ssout,"",1) == -1) die_write();
       }
     }

   if (len >= ADDR) die(11);
  }

 if (substdio_flush(&ssout) == -1) die_write();
 if (fsync(intdfd) == -1) die_write();

 if (link(intdfn,todofn) == -1) die(66);

 triggerpull();
 die(0);
}

int tapcheck(t)
char t;
{
  int i = 0;
  int j = 0;
  int x = 0;
  int negate = 0;
  stralloc curregex = {0};
  char tmpbuf[200];

  while (j < tap.len) {
    i = j;
    if ( tap.s[i]==t || tap.s[i]=='A'){
            while ((tap.s[i] != ':') && (i < tap.len)) i++;
            i++;
            j=i;
            while ((tap.s[i] != ':') && (i < tap.len)) i++;
            if (tap.s[j] == '!') {
              negate = 1;
              j++;
            }
            stralloc_copys(&tapaddr, &tap.s[i+1]);

            stralloc_copyb(&curregex,tap.s + j,(i - j));
            stralloc_0(&curregex);
            x = matchregex(chkaddr.s, curregex.s, tmpbuf);


            if ((negate) && (x == 0)) {
              return 1;
            }
            if (!(negate) && (x > 0)) {
              return 1;
            }
    }
    while ((tap.s[i] != '\0') && (i < tap.len)) i++;
    j = i + 1;
    negate = 0;


  }
  return 0;
}
